src/librustc/ty/structural_impls.rs

```diff
@@ -839,14 +839,17 @@ impl<'tcx> TypeFoldable<'tcx> for ty::ExistentialProjection<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for ty::ProjectionTy<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
+        let folded_trait_ref = self.trait_ref(folder.tcx()).fold_with(folder);
```

Comment:
Can folding a trait ref be overloaded? i.e. is there any `fold_trait_ref`? If not you can just fold substitutions.